src/v/archival/archiver_operations_impl.cc

```diff
@@ -327,9 +327,117 @@ class archiver_operations_impl : public archiver_operations_api {
+        // Generate archival metadata for every uploaded segment and
+        // apply it in offset order.
+        // Stop on first error.
+        // Validate consistency.
+        try {
+            size_t num_segments = upl_res.results.size();
+            if (
+              num_segments != upl_res.stats.size()
+              || num_segments != upl_res.metadata.size()) {
+                vlog(
+                  _rtclog.error,
+                  "Bad 'admit_uploads' input. Number of segments: {}, number "
+                  "of "
+                  "stats: {}, number of metadata records: {}",
+                  num_segments,
+                  upl_res.stats.size(),
+                  upl_res.metadata.size());
+                co_return error_outcome::unexpected_failure;
+            }
+            bool validation_required
+              = !config::shard_local_cfg()
+                   .cloud_storage_disable_upload_consistency_checks();
+            auto part = _pm->get_partition(upl_res.ntp);
+            std::vector<cloud_storage::segment_meta> metadata;
+            for (size_t ix = 0; ix < num_segments; ix++) {
+                auto sg = upl_res.results.at(ix);
+                auto st = upl_res.stats.at(ix);
+                auto meta = upl_res.metadata.at(ix);
+
+                if (sg != upload_result::success) {
+                    break;
+                }
+                // validate meta against the record stats
+                if (st.has_value() && validation_required) {
+                    if (!this->segment_meta_matches_stats(
+                          meta, st.value(), _rtclog)) {
+                        break;
+                    }
+                } else {
+                    vlog(
+                      _rtclog.debug,
+                      "Segment self-validation skipped, meta: {}, record stats "
+                      "available: {}, validation_required: {}",
+                      meta,
+                      st.has_value(),
+                      validation_required);
+                }
+                metadata.push_back(meta);
+            }
+            // optionally validate the data
+            bool is_validated = false;
+            if (validation_required) {
+                auto num_accepted = part->manifest().safe_segment_meta_to_add(
+                  metadata);
+                if (num_accepted == 0) {
+                    vlog(
+                      _rtclog.error,
+                      "Metadata can't be replicated because of the validation "
+                      "error");
+                    co_return cloud_storage::error_outcome::failure;
+                } else if (num_accepted < metadata.size()) {
+                    vlog(
+                      _rtclog.warn,
+                      "Only {} segments can't be admitted, segments: {}",
+                      num_accepted,
+                      metadata);
+                    metadata.resize(num_accepted);
```

Comment:
nit: this seems like a brittle interface with safe_segment_meta_to_add. I guess it makes sense if we only consider contiguous uploads, and the first few uploads succeed. Are we expecting compacted reuploads to be running in a different admission of uploads?